src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -129,58 +135,87 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
+        }
 
-            checksums.Clear();
+        async Task SynchronizeProjectAssetCollectionAsync<TAsset>(Func<ProjectStateChecksums, ChecksumCollection> getChecksums, CancellationToken cancellationToken)
+        {
+            await Task.Yield();
+            using var _ = PooledHashSet<Checksum>.GetInstance(out var checksums);
 
-            // Then synchronize the info about all the documents within.
-            await CollectChecksumChildrenAsync(checksums, projectChecksums.Documents).ConfigureAwait(false);
-            await CollectChecksumChildrenAsync(checksums, projectChecksums.AdditionalDocuments).ConfigureAwait(false);
-            await CollectChecksumChildrenAsync(checksums, projectChecksums.AnalyzerConfigDocuments).ConfigureAwait(false);
+            foreach (var projectChecksums in allProjectChecksums)
+                AddAll(checksums, getChecksums(projectChecksums));
 
-            await this.SynchronizeAssetsAsync<object, VoidResult>(
-                assetPath: AssetPath.ProjectAndDocuments(projectChecksums.ProjectId), checksums, callback: null, arg: default, cancellationToken).ConfigureAwait(false);
+            await this.SynchronizeAssetsAsync<TAsset, VoidResult>(
+                AssetPath.SolutionAndTopLevelProjectsOnly, checksums, callback: null, arg: default, cancellationToken).ConfigureAwait(false);
         }
 
-        async ValueTask CollectChecksumChildrenAsync(HashSet<Checksum> checksums, ChecksumsAndIds<DocumentId> collection)
+        async Task SynchronizeProjectDocumentsAsync(ProjectStateChecksums projectChecksums, CancellationToken cancellationToken)
         {
-            // This GetAssetsAsync call should be fast since they were just retrieved above.  There's a small chance
-            // the asset-cache GC pass may have cleaned them up, but that should be exceedingly rare.
-            var allDocChecksums = await this.GetAssetsAsync<DocumentStateChecksums>(
-                AssetPath.ProjectAndDocuments(projectChecksums.ProjectId), collection.Checksums, cancellationToken).ConfigureAwait(false);
-            foreach (var docChecksums in allDocChecksums)
+            await Task.Yield();
+            using var _1 = PooledHashSet<Checksum>.GetInstance(out var checksums);
+
+            AddAll(checksums, projectChecksums.Documents.Checksums);
+            AddAll(checksums, projectChecksums.AdditionalDocuments.Checksums);
+            AddAll(checksums, projectChecksums.AnalyzerConfigDocuments.Checksums);
+
+            // First, fetch all the DocumentStateChecksums for all the documents in the project.
+            using var _2 = ArrayBuilder<DocumentStateChecksums>.GetInstance(out var allDocumentStateChecksums);
+            await this.SynchronizeAssetsAsync<DocumentStateChecksums, ArrayBuilder<DocumentStateChecksums>>(
+                assetPath: AssetPath.ProjectAndDocuments(projectChecksums.ProjectId), checksums,
```

Comment:
>ProjectAndDocuments [](http://example.com/codeflow?start=37&length=19)

could this just use AssetPathKind.Documents?

Reply:
yes.  that seems appropriate.  thanks!